### PR TITLE
IA3 adaptors

### PR DESCRIPTION
**Pull request by jagnusson**
Changes proposed in this pull request:

- Adds a function to modify a Hugging Face transformer with IA3 adaptors

## Results on piqa
A related [PR in catwalk](https://github.com/allenai/catwalk/pull/81) implements an example of how these adaptors can be trained. While hardly impressive results, the IA3 implementation manages to reduce validation loss and recover much of the accuracy of the fully tuned equivalent for all the architectures for which default configurations are provided. The `gpt-j-6b` full tune is not able to run on a single gpu while the IA3 training is able to fit due to having far fewer optimizer states for its fewer trainable parameters.

Screen Shot 2022-09-13 at 6 57 54 PM

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/tango/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/tango/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [x] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: tango/integrations/transformers/ia3.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        size_out = x.size()[:-1] + (self.nf,)
+        x = torch.addmm(self.bias, x.view(-1, x.size(-1)), self.weight)
+        x = x.view(size_out)  # ... * self.nf
+
+        ia3_params = getattr(self, self.ia3_param_names)
+
+        if ia3_params.requires_grad:
+            if self.unfuse_size is not None:
+                # non_q means k and v
+                q, non_q = x[:, :, : self.unfuse_size], x[:, :, self.unfuse_size :]
```
Comment (AkshitaB) [inline tango/integrations/transformers/ia3.py:152]: Which transformer models have fused `Conv1D` layers? 

It's minor, but we can probably refactor away the code repetition.

File: tango/integrations/transformers/ia3.py [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from transformers import PreTrainedModel
+from transformers.modeling_utils import Conv1D
+
+
+@dataclass
+class WithIA3Config:
```
Comment (AkshitaB) [inline tango/integrations/transformers/ia3.py:13]: Docstring will be helpful.

File: tango/integrations/transformers/ia3.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    "facebook/opt-6.7b": OPT_IA3_CONFIG,
+    "facebook/opt-13b": OPT_IA3_CONFIG,
+    "facebook/opt-30b": OPT_IA3_CONFIG,
+    "facebook/opt-66b": OPT_IA3_CONFIG,
+    "EleutherAI/gpt-j-6B": GPT_J_IA3_CONFIG,
+}
+
+
+class LinearWithIA3(nn.Module):
+    def __init__(self, linear_layer, ia3_param_names, unfuse_size: int = None):
```
Comment (AkshitaB) [inline tango/integrations/transformers/ia3.py:82]: Docstring should be added. It took me a moment to understand that `unfuse_size` was the `hidden_size`.

**Review by AkshitaB — CHANGES_REQUESTED** [review]
Only minor comments.

File: tango/integrations/transformers/ia3.py (hunk 1 above)
Comment (jagnusson) [reply tango/integrations/transformers/ia3.py:152]: Of the architectures I made configs for here, only gpt2 uses `Conv1d` layers. Its attention layer is fused but its mlp is not so unfortunately we need both the `if self.unfuse-size is not None:` and `else:` control flow paths here.

Meanwhile bloom uses `nn.Linear` layers and also has a fused attention layer.

I could factor out the unfusing code here from the two classes. Or did you have something else in mind?

Comment (AkshitaB) [reply tango/integrations/transformers/ia3.py]: Yes, that's what I had in mind.

File: tango/integrations/transformers/ia3.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    """
+    A function to add ia3 adaptors to the given transformer. Code modified from
+    `t-few <https://github.com/r-three/t-few/blob/217cfa3b73aa66a07594826e4ebbbc516b331461/src/models/lora.py>`_
+    and Qinyuan Ye
+
+
+    :param model:
+        A :class:`~transformers.PreTrainedModel` to modify.
+    :param config:
+        A :class:`~tango.integrations.transformers.ia3.WithIA3Config` that specifies the layers to modify.
```
Comment (dirkgr) [inline tango/integrations/transformers/ia3.py:207]: Is there any chance we could automatically detect the right config, at least in some cases?

Comment (jagnusson) [reply tango/integrations/transformers/ia3.py:207]: So we could make it look up the known configs in `MODEL_NAME_TO_CONFIG` in [tango/integrations/transformers/ia3.py](https://github.com/allenai/tango/pull/403/files/68db503f1da57b4c1f8498819c084158badb8830#diff-f3d77c73a5bc3a8ac400e155fd8bbcd531b7b93c71759fb487d22ac20ea10425).

But if you mean trying to figure out a config from scratch just based on looking at the model architecture, that might be pretty difficult. Just from the few that we support right now they all have very different names for the layers we need. And there's enough variation in how the nodes can be nested that we can't find the layers we need by just looking for a Linear layer with a certain position in the model graph.

Comment (dirkgr) [reply tango/integrations/transformers/ia3.py:207]: No, I mean, can we just find the name of the model given the model, and do the lookup that way?

Comment (dirkgr) [reply tango/integrations/transformers/ia3.py:207]: I checked: We can look up `transformer_model.config.name_or_path`. If it matches anything in the dictionary of configs, we can use it automatically.

Comment (jagnusson) [reply tango/integrations/transformers/ia3.py:210]: Yup working on a commit to do that!

Comment (jagnusson) [reply tango/integrations/transformers/ia3.py:210]: Okay it's working now and I also updated the Catwalk end to use this: https://github.com/allenai/catwalk/pull/81/commits/9800e1249f5e311b16429635bfa957da6f3fd184

**Review by dirkgr — CHANGES_REQUESTED** [review]

File: tests/integrations/transformers/ia3_test.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+def test_ia3():
+
+    config = GPT_2_IA3_CONFIG
+    model_name = "sshleifer/tiny-gpt2"
+
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+
+    input_seq = tokenizer(["A tiny test on a tiny model."], return_tensors="pt")
+
+    model = AutoModelForCausalLM.from_pretrained(model_name)
```
[the remark this answers is not in the document]
Comment (jagnusson) [reply tests/integrations/transformers/ia3_test.py:16]: Oops good catch!

**Review by dirkgr — APPROVED** [review]
